### PR TITLE
feat(kubernetes): Add styling based on current context

### PR DESCRIPTION
**Pull request by jankatins**
#### Description
This PR implements the suggestion from https://github.com/starship/starship/issues/570 to add an ability to customize the configuration of the kubernetes module style, based on the current context.

A new variable is added to the config section, called environments, which is a list of possible customizations. Each such customization is an object with a context_pattern regex, which matches context name, and an optional style and icon that will override the global configuration, if the currently used context matched the context_pattern.

Based on multiple attempts to add per-context styling and symbols to the kubernetes module.

- https://github.com/starship/starship/pull/1568 by @lht  -> base (rebased and adjusted to current HEAD)
- https://github.com/starship/starship/pull/614 by @nomaed -> naming, symbol, some tests

Rebase, combined, and extended by @jankatins

#### Motivation and Context
It is common to work with development, testing, staging and production environments when deploying to kubernetes clusters, especially with continuous deployment. However, it is too easy to be using a production context (for example, for inspecting logs) and forgetting to switch to the personal development context when deploying any changes.

This change helps giving better visibility, by allowing to change the style of the kubernetes module in the prompt to draw attention, or changing the color to red (as an example) if using a context that should be used with care.

Closes: https://github.com/starship/starship/issues/570 (original request)
Closes: https://github.com/starship/starship/pull/1568 (superseded PR)
Closes: https://github.com/starship/starship/pull/614 (superseded PR)

#### Screenshots (if appropriate):

image

#### How Has This Been Tested?

- [x] I have tested using **MacOS**
- [ ] I have tested using **Linux**
- [ ] I have tested using **Windows**

#### Checklist:

- [x] I have updated the documentation accordingly.
- [x] I have updated the tests accordingly.

### CODE REVIEW

File: src/modules/kubernetes.rs
Comment (jankatins) [inline src/modules/kubernetes.rs:1103]: This misses a test for the warning, but my rust (beginner...) and google knowledge didn't find a possibility without adding another crate.

File: src/modules/kubernetes.rs
Comment (jankatins) [inline src/modules/kubernetes.rs:984]: https://github.com/starship/starship/pull/614 contains more tests, I tried to port the most relevant over. Not sure if I should port more?

**Comment by jankatins** [comment]
I guess https://github.com/starship/starship/pull/4505/files should be reworked to add a 'ignore' field to an environment? One could argue that the context alias should also be similarly be reworked as a per alias defined in the environment?

CC: @DanielKneipp @davidkna

**Comment by jankatins** [comment]
The [windows failure](https://github.com/starship/starship/actions/runs/3357644660/jobs/5563604803) seems to be in an unrelated test:

```
failures:
    modules::git_status::tests::doesnt_show_ignored_file
```

File: docs/config/README.md
Comment (jankatins) [inline docs/config/README.md:2250]: Alternative would be "contexts" (and renaming `context_pattern` to `pattern`)

**Comment by jankatins** [comment]
Rebased on top of current master

**Comment by jankatins** [comment]
Clippy failures seem unrelated, as the problems are in other modules.

**Comment by jankatins** [comment]
Rebased and, wohoo, green builds!

File: src/configs/kubernetes.rs
Comment (chipbuster) [inline src/configs/kubernetes.rs:41]: I am not as familiar with schemars/serde as I'd like to be, so I'd like a quick check from another maintainer on this part to make sure the macros on this struct look correct.

File: src/modules/kubernetes.rs
Comment (chipbuster) [inline src/modules/kubernetes.rs:985]: This set of tests looks good for the change. A number of the tests from #614 seem to be covered by other existing tests at this point.

**Review by chipbuster — COMMENTED** [review]
The current PR looks good to me, though we should work out what we want to do with naming for the config options before we merge this, since that will be difficult to change later on.

I would also like another reviewer to check the macros on the new `KubernetesEvnironmentConfig` struct, since I'm not as familiar with schemars as I'd like to be.

As far as the proposed changes w.r.t putting alises into the environment config itself,  we can either include it in this PR, or merge this (after resolving naming) and have a new PR that includes that functionality. I will leave it to @jankatins to decide which they prefer.

File: docs/config/README.md
Comment (chipbuster) [reply docs/config/README.md:2285]: @jankatins I don't have lots of experience with Kubernetes, but some searching suggests that "contexts" is the less ambiguous (and more idiomatic) term for Kubernetes. Does that square with your experience? If so, I would vote for going with contexts and rename the pattern variable. The term "environment" is overloaded enough as it is.

Would also be interested in @nomaed's opinion on this.

As far as the aliases go, I think that sounds like a good idea (but again, I don't know much about Kubernetes). It seems to me that it basically eliminates the need for multiple sets regexes, which are a little fiddly to write. In the worst case, we can deprecate the user_alias and context_alias options now, then start emitting warnings for anyone using them in a future version, before finally getting rid of them for 2.0.

File: src/configs/kubernetes.rs
Comment (davidkna) [reply src/configs/kubernetes.rs:41]: The macros look correct to me.

File: docs/config/README.md
Comment (chipbuster) [reply docs/config/README.md:2285]: #4505 also seems to suggest that "contexts" is the preferred term.

Comment (jankatins) [reply docs/config/README.md:2285]: So:
-Rename environment to context
- Add user_pattern, user_alias, context_alias (given that we have user and context pattern, I would keep the `context_` prefix to `context_pattern` -> user_pattern would only be evaluated if set.
- Maker that the default way to set user and context aliases: if not set via the matched context
- look at the old way and omit a warning
- If can, also port the "set alias from regex group" feature.

**Comment by jankatins** [comment]
- [x] Renamed "environments" to "contexts" and added the other three config items
- [x] Added context matching based on user_pattern (default: no matching happens, context_pattern alone is responsible for matching a context)
- [x] Added context_alias and user_alias (and deprecated the old way by removing the docs for it :-)

This has one regression: before it was possible to define a match group in the regex and then use that in the alias. I haven't implemented that.

File: docs/config/README.md
Comment (jankatins) [reply docs/config/README.md]: Done now, apart from the regex group feature

**Comment by jankatins** [comment]
Rebased... From my perspective it's ready for review again...

File: src/modules/kubernetes.rs
Comment (chipbuster) [inline src/modules/kubernetes.rs:204]: I'm find the computation of this variable's value hard to read. I don't know if the fact that I'm a little tired right now affects it, but there's a lot of error handling, logging, short-circuiting, and nested things mixed in here.

I will revisit this tomorrow and see if I still have problems with it (and if so, hopefully have some suggestions).

**Review by chipbuster — COMMENTED** [review]
I'm not 100% happy with the body of the module function right now, but I'm also pretty fatigued. Let me get some sleep and get back to you on that.

Everything else looks alright to me, except for the loss of capture groups in the regexes. Is that not implemented just because of additional complexity, or is there some technical blocker that would prevent us from implementing that in the future?

I'm wary of deprecating existing functionality without having a suitable replacement or workaround ready, so if we aren't 100% certain that this can be written without too much hassle, I'm going to be a little nervous about merging this.

File: src/modules/kubernetes.rs
Comment (jankatins) [reply src/modules/kubernetes.rs:204]: Getting a better codestyle here (and also below with all the maps there) would be awesome. I'm a rust beginner, so in some places, I basically changed the code until it compiled and worked :-(

File: src/modules/kubernetes.rs
Comment (jankatins) [inline src/modules/kubernetes.rs:291]: I still have no idea why I need clone here, but that was the only way I could get this whole lines to compile :-(

**Comment by jankatins** [comment]
> I'm wary of deprecating existing functionality without having a suitable replacement or workaround ready, so if we aren't 100% certain that this can be written without too much hassle, I'm going to be a little nervous about merging this.

So right now my problem with this feature is that it now would need two times running the regex: ones during the map which needs to return the bool to select an entry and the second time to get a match entry to fill out the context alias. Not sure if that is a performance penalty or not. If not, It's simply rerunning the regex.

File: src/modules/kubernetes.rs
Comment (chipbuster) [reply src/modules/kubernetes.rs:204]: With a little bit of fiddling, I've been able to get this down to this:

```
       let is_context_match = match_regex(&kube_ctx, &context_config.context_pattern);
        if !is_context_match {
            return false;
        };
        if let Some(user_pattern) = context_config.user_pattern {
            if let Some(user) = kube_user.as_ref() {
                return match_regex(user, user_pattern);
            } else {
                // If there is a user pattern but no user, the pattern cannot match
                return false;
            }
        } else {
            // If there is no user pattern, then this context config matches
            return true;
        }
```

with the following definition of `match_regex`:
```
fn match_regex(target: &str, pattern: &str) -> bool {
    match regex::Regex::new(pattern) {
        Ok(re) => re.is_match(&target),
        Err(error) => {
            log::warn!("Could not compile regex pattern `{}`:\n{}", pattern, error);
            false
        }
    }
}
```

I'm still not entirely happy with this, because:

1. The `match_regex` name is terrible.
2. I think the body of the closure in `find` can still be cleaned up a little bit
3. I think we can modify match_regex to do capture group parsing and output in a single pass instead of having to do two passes

but I think it's a starting point.

Comment (chipbuster) [reply src/modules/kubernetes.rs:204]: One downside here is that the error messages won't contain the context of whether it's a context-regex or a user-regex, but I doubt most users will have the same regex for both (and if they do, they both need to be fixed).

File: src/modules/kubernetes.rs
Comment (chipbuster) [inline src/modules/kubernetes.rs:176]: Something I just realized while poking around this module: it looks like in all subsequent uses of the `ctx_components` variable, we iterate over the components, check if they're `Some`, and then check if one of their components is a `Some`.

Why don't we just filter the vector here to only contain non-`None` components?

Comment (jankatins) [reply src/modules/kubernetes.rs:176]: you mean add a `.filter_map(|x| x)` here to filter out the `None` and declare it as `Vec<KubeCtxComponents>`?

Comment (chipbuster) [reply src/modules/kubernetes.rs:176]: On further reflection, I think we can change this `map` to `filter_map` and get the same result.

**Comment by chipbuster** [comment]
@jankatins I might have a few spare hours this week to play around with this. Would you be okay with me applying some of the changes I've suggested here?

**Comment by jankatins** [comment]
@chipbuster Please go ahead: I'm looking forward to see proper rust code and not my trial and error way of coding :⁠-⁠)

**Comment by chipbuster** [comment]
@jankatins I've made some of the changes we discussed. Among the changes:

- Move the alias functions to a `deprecated` submodule, instead of changing their names.
- Filter out all `None` ContextComponents to greatly simplify processing of components
- Create a regex-matching function to simplify a lot of matching logic
- Make Clippy upset (whoops, should fix that!)

However, I'm not really a Kubernetes person, and I still don't understand a few of the things that are happening here.In particular, we currently

**Comment by chipbuster** [comment]
@jankatins On further reflection, I think the nesting of options in the module makes it more confusing than it needs to be (it's probably also why I had trouble reading the PR initially and why you had to do a bit of compiler-driven type-tetris to get this thing building.

The changes I made in the previous commit do help this somewhat, but it also bloats the PR. I've made a separate PR with my cleanup changes (#4743) and assuming that it gets merged, I'll rebase your changes on top of that and we can go from there.

File: src/modules/kubernetes.rs
Comment (jankatins) [inline src/modules/kubernetes.rs:131]: The more I stare at this, the more this feels wrong: IMO this should end up in a `<Option<KubeCtxComponents>`  -> either the context is found in the contexts or not:

```yaml
current-context: work
contexts:
- context:
    cluster: work
    namespace: work-dev
    user: clusterUser
  name: 50hzk8s
- context:
    cluster: docker-desktop
    user: docker-desktop
  name: docker-desktop
```

i.e in `get_kube_ctx_component` we filter for the name, so there will at most be one KubeCtxComponents coming out of this, why should this become a vector...?

```rust
fn get_kube_ctx_component(filename: path::PathBuf, current_ctx: &str) -> Option<KubeCtxComponents> {
...
    let ctx_yaml = conf["contexts"].as_vec().and_then(|contexts| {
        contexts
            .iter()
            .filter_map(|ctx| Some((ctx, ctx["name"].as_str()?)))
            .find(|(_, name)| *name == current_ctx)  // <- HERE
    });

    let ctx_components = KubeCtxComponents {
...
    };

    Some(ctx_components)
}
```

And if it is None, the whole module should be aborted, IMO...

Comment (jankatins) [reply src/modules/kubernetes.rs:131]: E.g. this seems to have introduced the vec: https://github.com/starship/starship/pull/3569/files#diff-893760ffe78f91e7934d6da73675dbc83b63e62bbce7341f70310b0ab1187442L89

**Comment by jankatins** [comment]
I can do the rebase, but I would probably squash all commits down and do it then as the rebase of all commits looks ... painful.

I'm not sure if I can get to it before xmas, though. Might be that I have some time tomorrow in the train, but otherwise only after...

**Comment by jankatins** [comment]
@chipbuster I squash-rebased (added all authors as coauthors) and added a commit which removes some options, switches to `<Option<KubeCtxComponents>`, and renames some functions to make it more clearer what is actually going on (thats at least what I though...). 

This still doesn't add the match group feature back :-(

File: src/modules/kubernetes.rs
Comment (davidkna) [reply src/modules/kubernetes.rs:234]:
```suggestion
                "user" => display_user.as_deref().map(Cow::Borrowed).map(Ok),
```

**Comment by davidkna** [comment]
I think some of the matches in this module can be simplified via [`let` `else` statements](https://blog.rust-lang.org/2022/11/03/Rust-1.65.0.html#let-else-statements).

File: src/modules/kubernetes.rs
Comment (chipbuster) [reply src/modules/kubernetes.rs]: > in get_kube_ctx_component we filter for the name, so there will at most be one KubeCtxComponents coming out of this, why should this become a vector...?

The only condition I can think of that would cause this is if there are multiple Kubernetes configs that define the same context. If that isn't just a hard error, then you need some way to resolve conflicting settings.

Is this allowed by Kubernetes at all? If so, what's the correct behavior?

Comment (chipbuster) [reply src/modules/kubernetes.rs]: I think I understand now. The logic in using a vector of components is to implicitly (throughout the rest of the module) implement the merging logic detailed at https://kubernetes.io/docs/concepts/configuration/organize-cluster-access-kubeconfig/#merging-kubeconfig-files

I'm not sure I'm a fan of having things spread out like this. Perhaps the move is to put the logic for implementing the merge procedure into `get_kube_ctx_components`? Right now, I think if you have incomplete context components e.g. one config that only defines a user, and another config that only defined the namespace, the module will incorrectly only pick up on the first one defined.

Also surprisingly, it seems we don't actually have a test case for config merging.

Comment (jankatins) [reply src/modules/kubernetes.rs]: I actually think it's now correct: 

> Never change the value or map key. Example: Preserve the context of the first file to set current-context. Example: If two files specify a red-user, use only values from the first file's red-user. Even if the second file has non-conflicting entries under red-user, discard them.

My reading is, that the last sentence implies that the first context found is used for everything, no matter if another config defines "non-conflicting" items (eg. your example: "one config that only defines a user, and another config that only defined the namespace").

Comment (chipbuster) [reply src/modules/kubernetes.rs]: Huh, I guess that's right. (And we did get [a comment](https://github.com/starship/starship/pull/3569#issuecomment-1102130317) that it wasn't working properly, though I don't really know what the issue was).

Is there a way we can add a test for this?

Comment (jankatins) [reply src/modules/kubernetes.rs]: Ok, just verified this behaviour: if there is already a context with that name, only the config from that file is used and not "added" or overwritten from a second file: both ` one.yml` and `two.yml` contain a context with `name: work`, but the values only come from the context which was defined first:

```
λ  KUBECONFIG=$(pwd)/one.yml:$(pwd)/two.yml kubectl config view # no "namespace" in one.yml, but in two.yml
apiVersion: v1
clusters: null
contexts:
- context:
    cluster: docker-desktop
    user: docker-desktop
  name: docker-desktop
- context:
    cluster: work-cluster1
    user: clusterUser1
  name: work
current-context: work
kind: Config
preferences: {}
users: null

λ  KUBECONFIG=$(pwd)/two.yml:$(pwd)/one.yml kubectl config view # no user in two.yml -> gets set to an empty string!
apiVersion: v1
clusters: null
contexts:
- context:
    cluster: docker-desktop
    user: docker-desktop
  name: docker-desktop
- context:
    cluster: work-cluster2
    namespace: work-dev2
    user: ""
  name: work
current-context: work
kind: Config
preferences: {}
users: null
```

**Comment by jankatins** [comment]
> I think some of the matches in this module can be simplified via [let else statements](https://blog.rust-lang.org/2022/11/03/Rust-1.65.0.html#let-else-statements).

@davidkna I tried to find some places where the pattern mapped, but in all the places I looked, we do not simple "unpack" a option, but do something to it, too. From my (beginner) understanding, I couldn't make it work with `let else` statements. Any chance you have an example how it would works?

Sorry :-(

**Comment by davidkna** [comment]
@jankatins I agree, on second look there aren't as many opportunities as I thought. Because the `else` branches often don't return or look at the error inside the result. There is some opportunity to improve in the deprecated functions, but that might not be worth the effort to refactor, i.e.:
```rust
    pub fn get_kube_user_alias<'a>(
        config: &'a KubernetesConfig,
        kube_user: &'a str,
    ) -> Cow<'a, str> {
        let Some(alias) = get_alias(&config.user_aliases, kube_user) else {
            return Cow::Borrowed(kube_user)
        }
        log::warn!("Usage of 'user_aliases' is deprecated and will be removed in 2.0; Use 'contexts' with 'user_alias' instead. (`{kube_user}` -> `{alias}`)");
        alias
    }

**Comment by jankatins** [comment]
> Edit: If pulled into a separate function, the context_config.user_pattern match could also benefit.

Fun, this worked even without a new function as it was in a closure...

Never looked at that part :-) Thanks, I learned something new :-)

**Comment by jankatins** [comment]
I added back the "use capture groups in alias" feature. for my taste I had to use a lot of `as_str/to_string` due to me and clippy having long conversations until it compiled and passed the tests...

@chipbuster @davidkna In my opinion it's ready for review again...

File: src/modules/kubernetes.rs
Comment (jankatins) [inline src/modules/kubernetes.rs:116]: This woul have been a nice let-else example, but I haven'T found a way to access the error (to print) in that form :-(

**Comment by jankatins** [comment]
@chipbuster @davidkna Rebased on latest master. Any chance to get a review to keep this going?

File: src/modules/kubernetes.rs
Comment (chipbuster) [inline src/modules/kubernetes.rs:89]: I suspect this is partially because I'm tired, but I've had a hell of a time trying to parse how this function is used. I've been staring at this and its call sites for almost 20 minutes now and there are still things I don't understand:

- Why do we take two `Option<&str>` and then an `Option<&String>`?
- What do these strings represent? (Every time I think I have it, one of them falls out of my head and I have to start thinking about it again).
- Why does the one of the call sites have a `Some(Some(something, _, _))` followed immediately by an `unwrap_or_else`?

I'm afraid I don't have a concrete suggestion right now (maybe will have one this weekend) but my gut is that this is too complex to merge as it is. Perhaps there's no better way to write it, in which case I'll grumble a little and approve, but I suspect there's a way to make this simpler than it currently is.

Comment (jankatins) [reply src/modules/kubernetes.rs:89]: I think it's more likely due to my inexperience with rust :-/

> Why do we take two Option<&str> and then an Option<&String>?

Mostly because the inputs were this and this version compiled :-/ I still haven't completely wrapped my head around when to use &str vs String vs &String...

> What do these strings represent? (Every time I think I have it, one of them falls out of my head and I have to start thinking about it again).

I think of this function as a wrapper around a regex search and replace: have a pattern, have a current value, have a replacement. The alias is the replacement. Only difference: if the pattern doesn't match, return None instead of current value.

> Why does the one of the call sites have a Some(Some(something, _, _)) followed immediately by an unwrap_or_else?

Again, probably because I got it compiled like that when using the filter_map (if you mean line 201) :-(

Just to clarify, the part around line 201:

```
   let (matched_context_config, display_context, display_user) = config.contexts.iter().find_map(|context_config| {
        let Some(context_alias) = get_aliased_name(Some(context_config.context_pattern), Some(&current_kube_ctx_name), context_config.context_alias) else {
            return None;
        };

        let user_alias = ... 
...
        Some((Some(context_config), context_alias, user_alias))
    }).unwrap_or_else(|| (None, current_kube_ctx_name.clone(), ctx_components.user));

```

So `Some(Some(something, _, _))` is actually `Some(Some(something), _, _)`,, so `Option<Tuple<Option<..>,..>> which hopefully is a bit more sane than an `Option<Option<Tuple>>`

>  but I suspect there's a way to make this simpler than it currently is.

I sure hope, but I'm currently out of my wits/experience for anything to make this easier. :-(

Comment (jankatins) [reply src/modules/kubernetes.rs:89]: Ok, came across this: https://rust-unofficial.github.io/patterns/idioms/coercion-arguments.html which argues that you should always go with `&str` instead of `&String`. Will make the adjustments after I understood that chapter...

Comment (jankatins) [reply src/modules/kubernetes.rs]: Did some refinement on the function... Hope it's clearer now. Also got rid of the `&String`

**Comment by jankatins** [comment]
I cleared up the alias function and renamed it to `replace_with_alias_if_pattern_matches()`. Hopefully this function is now clearer and a bit easier to understand.

**Comment by jankatins** [comment]
Rebased and force pushed on top of current master. 

@chipbuster I'm out of ideas what to improve here after the changes I did to the function which irritated you (see last commit in the PR). Any chance to get a review from you?

**Comment by jankatins** [comment]
@davidkna @chipbuster Rebased and resolved the conflict. any chance to get this reviewed?

**Comment by davidkna** [comment]
I've made an attempt to reduce the complexity a bit more, please have a look @jankatins: https://github.com/davidkna/starship/commit/ccfff3243a65dc1503b7bce69d1a7f20f90d2667

**Comment by jankatins** [comment]
@davidkna Wow, thank you, that commit was a great learning experience! Especially the pattern `let value = current_value?` was totally unknown to me.


I pulled it into my branch, added a `refactor(kubernetes):` prefix and pushed it. 

The windows test failure seems unrelated:

```
Error: Custom { kind: Uncategorized, error: PathError { path: "C:\\Users\\RUNNER~1\\AppData\\Local\\Temp\\.tmpTT07Ia", err: Os { code: 32, kind: Uncategorized, message: "The process cannot access the file because it is being used by another process." } } }


failures:
    modules::git_status::tests::doesnt_show_ignored_file

test result: FAILED. 994 passed; 1 failed; 0 ignored; 0 measured; 0 filtered out; finished in 48.27s
```

**Comment by jankatins** [comment]
Rebased and force pushed -> All tests are green

**Comment by jankatins** [comment]
Rebased and force pushed and all tests green: @davidkna @chipbuster Any chance to get this reviewed and over the finish line?

**Comment by jankatins** [comment]
@davidkna @chipbuster any chance to get this over the finish line?

**Comment by felipesere** [comment]
I would love to have this. @davidkna @chipbuster any way I can help get this merged? 💜

**Comment by jankatins** [comment]
@davidkna @chipbuster @matchai Any chance to get a review of the latest review fixes?

**Review by chipbuster — APPROVED** [review]
:+1: from me, this is looking great!

I'm very happy with this new interface for specifying context-specific styling. Assuming we don't encounter issues with it once it's out in the wild, we might consider porting this into other modules where people have been requesting changes, e.g. in some of the git modules.

**Comment by chipbuster** [comment]
And it looks like all our other contributors are listed as coauthors in at least one commit, so this PR should be ready to go.

@jankatins Thank you so much for the patience to stick it out on this PR!

**Comment by andytom** [comment]
Thank you for your contribution @jankatins and thank you reviewing @chipbuster.

**Pull request merged**
